src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,903 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "corpriv.h"
+#include "tailcallhelp.h"
+#include "dllimport.h"
+#include "formattype.h"
+#include "sigformat.h"
+#include "gcrefmap.h"
+#include "threads.h"
+
+#ifndef CROSSGEN_COMPILE
+
+FCIMPL2(void*, TailCallHelp::AllocTailCallArgBuffer, INT32 size, void* gcDesc)
+{
+    FCALL_CONTRACT;
+
+    _ASSERTE(size >= 0);
+
+    return GetThread()->AllocTailCallArgBuffer(static_cast<size_t>(size), gcDesc);
+}
+FCIMPLEND
+
+FCIMPL0(void, TailCallHelp::FreeTailCallArgBuffer)
+{
+    FCALL_CONTRACT;
+    GetThread()->FreeTailCallArgBuffer();
+}
+FCIMPLEND
+
+FCIMPL2(void*, TailCallHelp::GetTailCallInfo, void** retAddrSlot, void** retAddr)
+{
+    FCALL_CONTRACT;
+
+    Thread* thread = GetThread();
+
+    void** hijackedSlot;
+    void* hijackedRetAddr = thread->GetHijackedReturnAddress(&hijackedSlot);
```

Comment:
It seems that if you moved all of the return address extracting code into the GetHijackedReturnAddress (maybe renamed to GetReturnAddress), it would be simpler and easier to follow.

Suggestion:
```C++
    void* GetReturnAddress(void** retAddrSlot)
    {
#ifdef FEATURE_HIJACK
        if ((m_State & TS_Hijacked) && (retAddrSlot == m_ppvHJRetAddrPtr))
        {
            return m_pvHJRetAddr;
        }
#endif

        return *retAddrSlot;
    }
```

Reply:
Good idea, that is better.